sources/api/early-boot-config/src/provider/vmware.rs

```diff
@@ -125,23 +134,146 @@ impl VmwareDataProvider {
+                })?
+            }
+
+            UserDataEncoding::Raw => {
+                String::from_utf8(user_data_bytes).context(error::InvalidUtf8 {
+                    what: Self::GUESTINFO_USERDATA,
+                })?
+            }
+        };
+
+        let json = SettingsJson::from_toml_str(&user_data_string, "user data")?;
+        Ok(Some(json))
+    }
+
+    /// Request a key's value from guestinfo
+    // TODO: Figure out if it's possible to re-use the backdoor or ERPC channel (mutable
+    // references everywhere...)
+    fn backdoor_get_bytes(key: &str) -> Result<Option<Vec<u8>>> {
+        // Probe and access the VMWare backdoor.  `kernel lockdown(7)` may block "privileged"
+        // mode because of its use of `iopl()`.  According the the bug report below, in theory
+        // "privileged" mode is more reliable than "unprivileged" but both provide access to
+        // the same data so we fall back to "unprivileged" if the first call fails.
+        // https://github.com/lucab/vmw_backdoor-rs/issues/6
+        let mut backdoor = vmw_backdoor::probe_backdoor_privileged()
+            .or_else(|e| {
+                warn!(
+                    "Unable to access backdoor via privileged mode, using unprivileged: {}",
+                    e
+                );
+                vmw_backdoor::probe_backdoor()
+            })
+            .context(error::Backdoor {
+                op: "probe and acquire access",
+            })?;
+
+        let mut erpc = backdoor.open_enhanced_chan().context(error::Backdoor {
+            op: "open eRPC channel",
+        })?;
+
+        erpc.get_guestinfo(key.as_bytes())
+            .context(error::GuestInfo { what: key })
+    }
 }
 
 impl PlatformDataProvider for VmwareDataProvider {
     fn platform_data(&self) -> std::result::Result<Vec<SettingsJson>, Box<dyn std::error::Error>> {
         let mut output = Vec::new();
 
-        match Self::user_data() {
-            Err(e) => return Err(e).map_err(Into::into),
-            Ok(None) => warn!("No user data found."),
+        match Self::guestinfo_user_data() {
+            Err(e) => {
+                error!("Error retrieving user data from guestinfo: {}", e);
+            }
+            Ok(None) => warn!("No user data found via guestinfo"),
             Ok(Some(s)) => output.push(s),
         }
+
+        // If we didn't find any user data in guestinfo, try CD-ROM
+        if output.is_empty() {
+            match Self::cdrom_user_data() {
+                Err(e) => return Err(e).map_err(Into::into),
+                Ok(None) => warn!("No user data found."),
+                Ok(Some(s)) => output.push(s),
+            }
+        }
+
         Ok(output)
     }
 }
 
 // =^..^=   =^..^=   =^..^=   =^..^=
 
+// Acceptable user data encodings
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum UserDataEncoding {
+    #[serde(alias = "b64")]
+    #[serde(alias = "base64")]
+    Base64,
```

Comment:
I'm curious about the alias of "base64"; does that mean "Base64" and "base64" would both be accepted?  Should it be a rename instead of an alias, so only "base64" is accepted, or does VMware really allow those two alternate capitalizations?  I'm not sure how it interacts with `rename_all = "lowercase"` above - I would have expected that you wouldn't have needed either one, alias or rename, since the rename_all seems like it'd do the job on its own.

(Same question for the gzip version)

Reply:
The cloud-init VMWare data provider accepts `base64` and `b64`, and the same goes for the `gzip` stuff.

I only added the `rename_all = lowercase` to be friendly and if a user had `Base64` we would rename it to `base64` and accept it.

Reply:
I understand the `b64` alias because it's a different name, but `base64`/`Base64` are the same other than capitalization, so it seems like the `rename_all` would take care of it for you without another alias here?  I'm wondering if line 266 can go, basically.  (and same for gzip)

Reply:
The `rename_all` happens first, so if a user entered `Base64` it would be lowercased to `base64`, and without the alias `base64` it wouldn't match any variants in the enum.

We could flip the logic around and remove the `rename_all` and just have aliases `base64` and `b64`.  That way we'll match the upper _and_ lowercase versions, as well as `b64`, but then we'd also need an alias for `B64`... haha :)

I think the way it is currently is safest and allows us to match everything with the least amount of `alias`es.

Reply:
Ok, I found that ^ wasn't true, `Base64` doesn't get accepted.  I'm going to remove the `rename_all` and add aliases.  That behavior feels the most straightforward to me and has the fewest surprises.

Reply:
@bcressey and I hacked around this a bit, and ended up removing the `rename_all` and going with aliases.  Once "case-insensitive" de/serialization is finalized, we should use that: https://github.com/serde-rs/serde/pull/1902 .  Added a comment to that effect.